### PR TITLE
little endian and bounds checking elimination optimizations

### PR DESCRIPTION
**Pull request by foubarre**
On my macbook Intel Core i7 this makes the Write and Checksum benchmarks at least 2x faster.

### CODE REVIEW

File: xxHash32/xxHash32.go [hunk 1]
```diff
@@ -79,35 +82,34 @@ func (xxh *xxHash) Write(input []byte) (int, error) {
 		xxh.bufused += len(input) - r
 
 		// fast rotl(13)
-		p32 := xxh.v1 + (uint32(xxh.buf[p+3])<<24|uint32(xxh.buf[p+2])<<16|uint32(xxh.buf[p+1])<<8|uint32(xxh.buf[p]))*prime32_2
+		p32 := xxh.v1 + u32(xxh.buf[:])*prime32_2
```
Comment (foubarre) [inline xxHash32/xxHash32.go]: go compiler does many interesting optimizations here:
- u32 function is inlined
- on little endian platforms uint32 is read directly from bytes
- bounds checking elimination is propagated inside u32
(tested with go 1.8 on an amd64 platform)

File: xxHash32/xxHash32.go [hunk 2]
```diff
@@ -79,35 +82,34 @@ func (xxh *xxHash) Write(input []byte) (int, error) {
...
-		p32 = xxh.v3 + (uint32(input[p+3])<<24|uint32(input[p+2])<<16|uint32(input[p+1])<<8|uint32(input[p]))*prime32_2
-		xxh.v3 = (p32<<13 | p32>>19) * prime32_1
-		p += 4
-		p32 = xxh.v4 + (uint32(input[p+3])<<24|uint32(input[p+2])<<16|uint32(input[p+1])<<8|uint32(input[p]))*prime32_2
-		xxh.v4 = (p32<<13 | p32>>19) * prime32_1
-		p += 4
+	if p > n-16 {
+		// Nothing to do
+	} else {
+		ptr := uintptr(unsafe.Pointer(&input[p]))
```
Comment (foubarre) [inline xxHash32/xxHash32.go]: use pointer arithmetics to prevent bounds checking to happen inside the loop

File: xxHash32/xxHash32.go [hunk 3]
```diff
@@ -203,3 +211,8 @@ func Checksum(input []byte, seed uint32) uint32 {
 
 	return h32
 }
+
+func u32(buf []byte) uint32 {
+	// go compiler recognizes this pattern and optimizes it on little endian platforms
```
Comment (foubarre) [inline xxHash32/xxHash32.go]: Initially I did the same thing using convoluted unsafe.Pointer casts and little endian autp-detection before noticing the compiler is able to detect this pattern.

**Comment by pierrec** [comment]
Hello,

I am all up for performance improvements, but I dont want to use unsafe for that: besides sometimes introducing subtle bugs, there are environments that do not allow it, like app engine.
To reduce bound checks, it may be worth trying some of the tricks described [here](http://www.tapirgames.com/blog/go-1.7-bce).

I like the function factorization as it makes the code clearer though.

**Comment by foubarre** [comment]
I know about the BCE tricks from tapir blog but still wasn't able to fully remove BC from hot loops  without resorting to unsafe. (not saying it's impossible, just that I didn't manage to do it ;-))
I can just remove unsafe for now, it will lose 10/15 % performance but will still be much faster than before.

**Comment by foubarre** [comment]
Thanks for your lz4/xxHash projects btw, I stumbled upon them when looking for a go lz4 library which is Reader/Writer compatible and non cgo to evaluate lz4 in a project and they're very well written and easy to work with!

**Comment by foubarre** [comment]
New safe version is in, Checksum is slower but I found more optimization in Write that makes it faster in the end
bench results:
master:
```
Benchmark_XXH32-8            	10000000	       196 ns/op	       0 B/op	       0 allocs/op
Benchmark_XXH32_Checksum-8   	10000000	       183 ns/op	       0 B/op	       0 allocs/op
Benchmark_XXH64-8            	10000000	       212 ns/op	       0 B/op	       0 allocs/op
Benchmark_XXH64_Checksum-8   	10000000	       189 ns/op	       0 B/op	       0 allocs/op
```
optim - unsafe:
```
Benchmark_XXH32-8            	20000000	        91.2 ns/op	       0 B/op	       0 allocs/op
Benchmark_XXH32_Checksum-8   	20000000	        69.3 ns/op	       0 B/op	       0 allocs/op
Benchmark_XXH64-8            	20000000	        99.1 ns/op	       0 B/op	       0 allocs/op
Benchmark_XXH64_Checksum-8   	20000000	        59.6 ns/op	       0 B/op	       0 allocs/op
```
optim - safe:
```
Benchmark_XXH32-8            	20000000	        89.5 ns/op	       0 B/op	       0 allocs/op
Benchmark_XXH32_Checksum-8   	20000000	        78.9 ns/op	       0 B/op	       0 allocs/op
Benchmark_XXH64-8            	20000000	        74.5 ns/op	       0 B/op	       0 allocs/op
Benchmark_XXH64_Checksum-8   	20000000	        58.6 ns/op	       0 B/op	       0 allocs/op
```

**Comment by pierrec** [comment]
Great work!

**Pull request merged**